yt/utilities/lib/image_utilities.pyx

```diff
@@ -36,6 +41,370 @@ def add_points_to_greyscale_image(
+                buffer_weight[jj1][kk2] +=  sw * dvx * (1 - dvy)
+
+            if jj2 < Nx and kk1 >= 0:
+                buffer[jj2][kk1]        += swq * (1 - dvx) * dvy
+                buffer_weight[jj2][kk1] +=  sw * (1 - dvx) * dvy
+
+            if jj2 < Nx and kk2 < Ny:
+                buffer[jj2][kk2]        += swq * (1 - dvx) * (1 - dvy)
+                buffer_weight[jj2][kk2] +=  sw * (1 - dvx) * (1 - dvy)
+
+@cython.boundscheck(False)
+@cython.cdivision(True)
+@cython.wraparound(False)
+def add_cells_to_image_offaxis(
+    *,
+    const np.float64_t[:, ::1] Xp,
+    const np.float64_t[::1] dXp,
+    const np.float64_t[::1] qty,
+    const np.float64_t[::1] weight,
+    const np.float64_t[:, :] rotation,
+    np.float64_t[:, ::1] buffer,
+    np.float64_t[:, ::1] buffer_weight,
+    const int Nx,
+    const int Ny,
+    const int Npix_min = 4,
+):
+    cdef np.ndarray[np.float64_t, ndim=1] center = np.array([0.5, 0.5, 0.5])
+    cdef np.float64_t w0 = 1 / sqrt(3.)
+    cdef int i, j, k
+
+    cdef np.ndarray[np.float64_t, ndim=1] a = np.array([1., 0, 0]) * w0
+    cdef np.ndarray[np.float64_t, ndim=1] b = np.array([0, 1., 0]) * w0
+    cdef np.ndarray[np.float64_t, ndim=1] c = np.array([0, 0, 1.]) * w0
+
+    a = np.dot(rotation, a)
+    b = np.dot(rotation, b)
+    c = np.dot(rotation, c)
+
+    cdef np.ndarray[np.float64_t, ndim=1] o = center - (a + b + c) / 2
+
+    cdef int Nsx, Nsy
+    cdef np.float64_t dx_max = np.max(dXp)
+    # The largest cell needs to be resolved by at least this number of pixels
+    Nsx = max(Npix_min, int(ceil(2 * dx_max * sqrt(3) * Nx)))
+    Nsy = max(Npix_min, int(ceil(2 * dx_max * sqrt(3) * Ny)))
+    cdef int max_depth = int(ceil(log2(max(Nsx, Nsy))))
+    cdef int depth
+
+    cdef np.ndarray[np.float64_t, ndim=3] stamp_arr = np.zeros((max_depth, Nsx, Nsy), dtype=float)
+    cdef np.ndarray[np.uint8_t, ndim=3] stamp_mask_arr = np.zeros((max_depth, Nsx, Nsy), dtype=np.uint8)
+    cdef np.float64_t[:, :, ::1] stamp = stamp_arr
+    cdef np.uint8_t[:, :, ::1] stamp_mask = stamp_mask_arr
+
+    # Precompute the mip
+    for depth in range(max_depth):
+        if depth == 0:
+            stamp[0, 0, 0] = 1
+            continue
+        direct_integrate_cube(
+            o,
+            a,
+            b,
+            c,
+            stamp_arr[depth, :, :],
+            stamp_mask_arr[depth, :, :],
+            min(1 << depth, Nsx),
+            min(1 << depth, Nsy),
+        )
+
+        stamp_arr[depth] /= np.sum(stamp[depth])
+
+    # Iterate over all cells, applying the stamp
+    cdef np.float64_t x, y, dx
+    cdef np.float64_t[:, ::1] rotation_view = np.ascontiguousarray(rotation)
+
+    cdef np.float64_t w, q, cell_max_width, sq3
+
+    sq3 = sqrt(3.)
+
+    # Local buffers
+    cdef np.float64_t **lbuffer
+    cdef np.float64_t **lbuffer_weight
+
+    cdef int num_particles = len(Xp)
+
+    with nogil, parallel():
+        lbuffer = <np.float64_t**> malloc(sizeof(np.float64_t*) * Nx)
```

Comment:
Is this a better solution than allocating the full `Nx*Ny` array?  If I am reading it right, this will still allocate the full thing on every processor, won't it?